etc/tune.c

```diff
@@ -2,141 +2,740 @@
+            i = 0;
+            /* Only the most basic checks */
+            errno = 0;
+            val = strtol(str, &endptr, base);
+            if ((val > INT_MAX || val < 0) || (errno != 0)) {
+               fprintf(stderr,"input #%d wrong\n", i+1);
+               exit(EXIT_FAILURE);
+            }
+            if (endptr == str) {
+               fprintf(stderr, "No input for #%d?\n", i+1);
+               exit(EXIT_FAILURE);
+            }
+            i++;
+            str = endptr + 1;
+            KARATSUBA_MUL_CUTOFF = (int)val;
+            errno = 0;
+            val = strtol(str, &endptr, base);
+            if ((val > INT_MAX || val < 0) || (errno != 0)) {
+               fprintf(stderr,"input #%d wrong\n", i+1);
+               exit(EXIT_FAILURE);
+            }
+            if (endptr == str) {
+               fprintf(stderr, "No input for #%d?\n", i+1);
+               exit(EXIT_FAILURE);
+            }
+            i++;
+            str = endptr + 1;
+            KARATSUBA_SQR_CUTOFF = (int)val;
+            errno = 0;
+            val = strtol(str, &endptr, base);
+            if ((val > INT_MAX || val < 0) || (errno != 0)) {
+               fprintf(stderr,"input #%d wrong\n", i+1);
+               exit(EXIT_FAILURE);
+            }
+            if (endptr == str) {
+               fprintf(stderr, "No input for #%d?\n", i+1);
+               exit(EXIT_FAILURE);
+            }
+            i++;
+            str = endptr + 1;
+            TOOM_MUL_CUTOFF = (int)val;
+            errno = 0;
+            val = strtol(str, &endptr, base);
+            if ((val > INT_MAX || val < 0) || (errno != 0)) {
+               fprintf(stderr,"input #%d wrong\n", i+1);
+               exit(EXIT_FAILURE);
+            }
+            if (endptr == str) {
+               fprintf(stderr, "No input for #%d?\n", i+1);
+               exit(EXIT_FAILURE);
+            }
+            i++;
+            str = endptr + 1;
+            TOOM_SQR_CUTOFF = (int)val;
+         case 'h':
+         default:
+            usage(argv[0]);
+            exit(EXIT_FAILURE);
+         }
+      }
    }
-   y = x;
 
-   for (x = 8; ; x += 2) {
-      t1 = time_sqr(x, 0);
-      t2 = time_sqr(x, 1);
-      printf("%d: %9llu %9llu, %9llu\n", x, t1, t2, t2 - t1);
-      if (t2 < t1) break;
+   s_raninit(seed);
+
+   ksm  = KARATSUBA_MUL_CUTOFF;
+   kss  = KARATSUBA_SQR_CUTOFF;
+   tc3m = TOOM_MUL_CUTOFF;
+   tc3s = TOOM_SQR_CUTOFF;
+
+   if ((bncore == 0) && (printpreset == 0)) {
+      /* Turn all limits from bncore.c to the max */
+      KARATSUBA_MUL_CUTOFF = INT_MAX;
+      KARATSUBA_SQR_CUTOFF = INT_MAX;
+      TOOM_MUL_CUTOFF = INT_MAX;
+      TOOM_SQR_CUTOFF = INT_MAX;
+#ifdef BN_S_MP_KARATSUBA_MUL_C
+      /*
+         The influence of the Comba multiplication cannot be
+         eradicated programmatically. It depends on the size
+         of the macro MP_WPARRAY in tommath.h which needs to
+         be changed manually (to 0 (zero)).
+       */
```

Comment:
Could we do something similar for MP_WPARRAY and define a tunable variable?

Reply:
Uhm&hellip;no, `MP_WARRAY` is not tunable itself and you cannot set the value to zero easily, that would need a change to how `MP_WARRAY` gets defined.

Reply:
Sure. My question is - can we make it tunable?

Reply:
`MP_WPARRAY`? As far as I understand it, the computed value is already at the optimum but I haven't checked the current literature yet.

My problem in `tune.c` was also a different one: I wanted to use the school-multiplication for checking the results, not the Comba multiplication.

> My question is - can we make it tunable?

Who's "we"? ;-)

Reply:
Generic "we" :) It doesn't mean it has to be done. It was just a question if warray would be a potential candidate for the tuner.